core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -350,7 +328,261 @@ public void run() {
+    @GuardedBy("this")
+    DecompressorRegistry decompressionRegistry;
+    @GuardedBy("this")
+    final LinkedList<InputStream> pendingMessages = new LinkedList<InputStream>();
+    @GuardedBy("this")
+    boolean pendingHalfClose;
+    @GuardedBy("this")
+    int pendingFlowControlRequests;
+    @GuardedBy("this")
+    boolean pendingFlush;
+
+
+    /**
+     * Get a transport and try to create a stream on it.
+     */
+    private final Runnable streamCreationTask = new Runnable() {
+      @Override
+      public void run() {
+        // Fill out timeout on the headers
+        headers.removeAll(TIMEOUT_KEY);
+        // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
+        // because timeout tolerates the clock difference between machines.
+        Long timeoutMicros = getRemainingTimeoutMicros();
+        if (timeoutMicros != null) {
+          if (timeoutMicros <= 0) {
+            maybeClosePrematurely(Status.DEADLINE_EXCEEDED);
+            return;
+          }
+          headers.put(TIMEOUT_KEY, timeoutMicros);
+        }
+
+        if (transportFuture.isDone()) {
+          ClientTransport transport;
+          try {
+            transport = transportFuture.get();
+          } catch (Exception e) {
+            maybeClosePrematurely(Status.fromThrowable(e));
+            return;
+          }
+          createStream(transport);
+        } else {
+          transportFuture.addListener(streamCreationTask, callExecutor);
+        }
+      }
+
+      @Override
+      public String toString() {
+        return "streamCreationTask";
+      }
+    };
+
+    DelayedStream(Metadata headers, Listener<RespT> callListener) {
+      this.headers = headers;
+      listener = new ClientStreamListenerImpl(callListener);
+      transportFuture = clientTransportProvider.get();
+      if (transportFuture == null) {
+        maybeClosePrematurely(Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+        return;
+      }
+      streamCreationTask.run();
+    }
+
+    /**
+     * Creates a stream on a presumably usable transport.
+     */
+    private void createStream(ClientTransport transport) {
+      synchronized (this) {
+        if (realStream == NOOP_CLIENT_STREAM) {
+          // Already cancelled
+          return;
+        }
+        Preconditions.checkState(realStream == null, "Stream already created: %s", realStream);
+        try {
+          realStream = transport.newStream(method, headers, listener);
+        } catch (IllegalStateException e) {
+          // The transport is already shut down. This may due to a race condition between Channel
+          // shutting down idle transports. Retry streamCreationTask, but Schedule it in an executor
+          // to avoid recursion.
+          callExecutor.execute(streamCreationTask);
+          return;
+        }
+        if (compressor != null) {
+          realStream.setCompressor(compressor);
+        }
+        if (decompressor != null) {
```

Comment:
So, there is a requirement on when `setDecompressor` should be called. Under such requirement, decompressor will never be set before `realStream` is created, thus we will never go into this branch. However, I like to keep `DelayedStream` unaware of such requirement, and just blindly pass down invocations. This will keep it simple and robust against any further contract changes.  

As long as the invocations to `DelayedStream.setDecompressor` meet such requirement, those to `realStream.setDecompressor` will do. In other words, `DelayedStream` will violate the contract on calling `realStream.setDecompressor` only if the caller of `DelayedStream.setDecompressor` does so.